### PR TITLE
Lock when writing to an SslStream

### LINKED ISSUE #417: The Write method cannot be called when another write operation is pending
* RabbitMQ version: 3.7.4
* Erlang version: 20.2.3
* Client library version: 5.1.0-rc1 (.NET Core)
* Windows Server 2016

We are getting these random write errors (sporadic) in an ASP.NET Core web app. Retrying the request after several seconds succeeds, but not sure why they're happening in the first place. We create a different connection per thread and create new models in each method that uses that connection. We also tried sharing a single long-lived connection but creating new models in each thread / method to no avail.

Message and StackTrace: System.NotSupportedException:  The Write method cannot be called when another write operation is pending.
   at System.Net.Security.SslStreamInternal.ProcessWrite(Byte[] buffer, Int32 offset, Int32 count, LazyAsyncResult asyncResult)
   at System.Net.Security.SslStream.Write(Byte[] buffer, Int32 offset, Int32 count)
   at RabbitMQ.Client.Impl.SessionBase.Transmit(Command cmd)
   at RabbitMQ.Client.Impl.ModelBase.ModelRpc(MethodBase method, ContentHeaderBase header, Byte[] body)
   at RabbitMQ.Client.Framing.Impl.Model._Private_ChannelOpen(String outOfBand)
   at RabbitMQ.Client.Framing.Impl.AutorecoveringConnection.CreateNonRecoveringModel()
   at RabbitMQ.Client.Framing.Impl.AutorecoveringConnection.CreateModel()

### PR DESCRIPTION
**Pull request by kjnilsson**
Network stream supports socket writes from multiple threads. SslStream
does not hence we need to lock before writing when using Ssl.

[#157088466]

May fix #417

### CODE REVIEW

File: projects/client/RabbitMQ.Client/src/client/messagepatterns/Subscription.cs [hunk 1]
```diff
@@ -238,7 +237,7 @@ public void Close()
                 {
                     if (Model.IsOpen)
                     {
-                        lock(m_basicCancelLock)
+                        lock(Model)
```
[the remark this answers is not in the document]
Comment (adamralph) [reply projects/client/RabbitMQ.Client/src/client/messagepatterns/Subscription.cs]: 👍 I'm not sure you should _ever_ take a lock on a public reference. It's deadlock territory.

[the remark this answers is not in the document]
Comment (kjnilsson) [reply projects/client/RabbitMQ.Client/src/client/messagepatterns/Subscription.cs]: Probably not a good idea as you say. We could lock on something inside the Model instead. I guess all rpcs inside a need to be locked over using a single lockobject per model as they cannot be pipelined.

**Comment by kjnilsson** [comment]
PR package: https://www.myget.org/feed/rabbitmq/package/nuget/RabbitMQ.Client/5.0.0-546ea94b

**Comment by kjnilsson** [comment]
This PR also includes locking around each RPC methods. As we cannot pipeline RPC calls we might as well lock around it. Combined with writing full frames this should increase model thread safety a fair bit.

**Pull request merged**